bentoml/server/bento_api_server.py

```diff
@@ -288,6 +288,41 @@ def setup_instruments(self):
             labelnames=['api_name', 'service_version'],
         )
 
+        def before_request():
+            self.request_start_time = default_timer()
+            api_name = _request_ctx_stack.top.request.url_rule.endpoint
+            self.metrics_request_in_progress.labels(
+                api_name=api_name, service_version=self.bento_service.version
+            ).inc()
+
+        def after_request(response):
+            api_name = _request_ctx_stack.top.request.url_rule.endpoint
+
+            # instrument request duration
+            total_time = max(default_timer() - self.request_start_time, 0)
+            self.metrics_request_duration.labels(
+                api_name=api_name,
+                service_version=self.bento_service.version,
+                http_response_code=response.status_code,
+            ).observe(total_time)
+
+            # instrument request total count
+            self.metrics_request_total.labels(
+                api_name=api_name,
+                service_version=self.bento_service.version,
+                http_response_code=response.status_code,
+            ).inc()
+
+            # instrument request in progress
+            self.metrics_request_in_progress.labels(
+                api_name=api_name, service_version=self.bento_service.version
+            ).inc()
+
+            return response
+
+        self.app.before_request_funcs = {None: [before_request]}
```

Comment:
I think this can be just `self.app.before_request(before_request)`?

Reply:
or
```
@self.app.before_request 
def before_request():
   ...
```

Reply:
Right. I was trying to make it the first one in the `before_request_funcs`, but think anyway it's rare users will touch this. I'll update.

Reply:
Ohh got it, I didn't realize that could be a problem.. I don't think BentoML user can have access to this?

Reply:
Most probably they won't notice this. 😄  Updated.